parlai/agents/transformer/modules/attention.py

```diff
@@ -224,9 +219,13 @@ def prepare_head(tensor):
             if static_kv:
                 mask = incr_state['prev_mask']
             else:
-                mask = torch.cat([incr_state['prev_mask'], mask], dim=2)
-                # Prepend along the key_len dimension (analogous to
-                # incr_state['prev_key'])
+                # Mask will be of size (B x query_len x key_len)
+                # During incremental decoding the query will only represent the next token,
+                # whereas the key/value will represent the entire sequence thus far.
+                # In such a case, we only want to look at the last element of the mask in the query dimension.
+                prev_mask = incr_state['prev_mask'][:, -query_len:, :]
```

Comment:
interesting, was this broken before?

Reply:
We never ran into this specific case before. This only breaks when 1) we are incremental decoding, and 2) sequences of different lengths are used between time steps.

When incremental decoding with decoder-only, the first step is the entire context. Then all following steps are of length one. So when transitioning from context to incremental decoding, `incr_state['prev_mask'].size(1)` will be the length of the context, but `mask.size(1)` will be one.